### PR TITLE
Fix worker exiting before all files are processed bug

### PR DESCRIPTION
**Pull request by omeshp**
- Fix worker exiting before all files are processed bug
- Mock fs in tests.

### CODE REVIEW

File: ItemLevelDownloader/Engine/worker.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { Logger } from './logger';
+export class Worker<T> {
+    private executionFunc: (item : T) => Promise<void>;
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Engine/worker.ts]: knit - rename to execute to be consistent with other names or name the others with Func suffix. I prefer the former.

File: ItemLevelDownloader/Engine/worker.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    async spawnWorker(resolve, reject) {
+        try {
+            await this.workerImplementation();
+            if (this.canExit()) {
+                Logger.logInfo("Exiting worker nothing more to process");
+                resolve();
+            }
+            else {
+                // spawn worker after 1 sec to check for items again.
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Engine/worker.ts]: add verbose log?

File: ItemLevelDownloader/Engine/worker.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    private getNextItem: () => T;
+    private canExit: () => boolean;
+
+    constructor(func: (item : T) => Promise<void>, getNextItem: () => T, canExit: () => boolean) {
+        this.executionFunc = func;
+        this.getNextItem = getNextItem;
+        this.canExit = canExit;
+    }
+
+    init(): Promise<void> {
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Engine/worker.ts]: have an id for worker and log it in messages to diagnose issues involving multiple workers?

File: ItemLevelDownloader/Engine/worker.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        }
+    }
+    
+    async workerImplementation() {
+        while (true) {
+            const item = this.getNextItem();
+            if (!item) {
+                break;
+            }
+
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Engine/worker.ts]: add verbose log?

File: ItemLevelDownloader/Engine/logger.ts [hunk 5]
```diff
@@ -4,42 +4,73 @@ import { ArtifactItemStore } from '../Store/artifactItemStore';
 
 export class Logger {
 
-    constructor(verbose: boolean) {
-        this.verbose = verbose
+    constructor(store: ArtifactItemStore, verbose: boolean) {
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Engine/logger.ts]: do we need to pass verbose to constructor?

File: ItemLevelDownloader/Engine/logger.ts [hunk 6]
```diff
@@ -4,42 +4,73 @@ import { ArtifactItemStore } from '../Store/artifactItemStore';
...
+                    + ", Skipped: " + skippedItems.length
+                    + ", Failed: " + failedItems.length
+                    + ", Time elapsed: " + ((currentTime.valueOf() - this.startTime.valueOf()) / 1000) + "secs");
+
+                if (this.store.itemsPendingProcessing()) {
+                    progressLogger();
+                }
+            }, 5000);
+
+            progressLogger();
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Engine/logger.ts]: will this be doing progress logging every 5 secs?

File: ItemLevelDownloader/Providers/filesystemProvider.ts [hunk 7]
```diff
@@ -39,12 +41,12 @@ export class FilesystemProvider implements models.IArtifactProvider {
...
             this.ensureDirectoryExistence(folder);
 
-            console.log("Downloading '%s' to '%s'", item.path, outputFilename);
+            Logger.logMessage('Downloading '+ item.path + ' to '+ outputFilename);
             const outputStream = fs.createWriteStream(outputFilename);
             stream.pipe(outputStream);
             stream.on("end",
                 () => {
-                    console.log(`Downloaded '${item.path}' to '${outputFilename}'`);
+                    Logger.logMessage(`Downloaded '${item.path}' to '${outputFilename}'`);
```
Comment (GitHubSriramB) [inline ItemLevelDownloader/Providers/filesystemProvider.ts]: good to be consistent with how you format the log string - use ${} everywhere or string concat... I think ${} looks cleaner...

**Review by GitHubSriramB — APPROVED** [review]

File: ItemLevelDownloader/Engine/logger.ts (hunk 5 above)
Comment (omeshp) [reply ItemLevelDownloader/Engine/logger.ts]: removed.

File: ItemLevelDownloader/Engine/logger.ts (hunk 6 above)
Comment (omeshp) [reply ItemLevelDownloader/Engine/logger.ts]: yes

File: ItemLevelDownloader/Engine/worker.ts (hunk 1 above)
Comment (omeshp) [reply ItemLevelDownloader/Engine/worker.ts]: done

**Pull request merged**